### ISSUE TITLE
An average quaternion cannot just be obtained by taking a Euclidean weighted mean

### ISSUE THREAD

**Issue by SoapJN** [opencv/opencv_contrib#2553]
https://github.com/opencv/opencv_contrib/blob/8fc6067b5987316d0586bc5db45c10aadad8c97f/modules/surface_matching/src/ppf_match_3d.cpp#L387-L401

The unit quaternions are no longer units when averaged.
We have two unit quaternions q1(w1, x1, y1, z1)、q2(w2, x2, y2, z2)，
||q1|| = sqr(w1) + sqr(x1) + sqr(y1) + sqr(z1) = 1,
||q2|| = sqr(w2) + sqr(x2) + sqr(y2) + sqr(z2) = 1.
Result of averaged the quaternions: q3((w1+w2) / 2, (x1+x2) / 2, (y1+y2) / 2, (z1+z2) / 2)
||q3|| = sqr((w1+w2) / 2) +  sqr((x1+x2) / 2) + sqr((y1+y2) / 2) + sqr((z1+z2) / 2) 
= 1/2 + w1*w2/2 +  x1*x2/2 + y1*y2/2 + z1*z2/2 
Only when q1 is equal to q2, the q3 is unit.
quaternions averaged cannot represent rigid transform matrix averaged

**Comment by SoapJN** [comment]
I studied one paper(**Averages of Rotations and Orientations in 3-space**) about this problem.  in the paper, it explicitly states that the method in the code is wrong

**Comment by catree** [comment]
[Averaging Quaternions](https://github.com/tolgabirdal/averaging_quaternions)?

**Comment by SoapJN** [comment]
> [Averaging Quaternions](https://github.com/tolgabirdal/averaging_quaternions)?

Your reference is correct ，This code have the problem

**Comment by SoapJN** [comment]
Your reference is correct ，This code can be optimized for higher accuracy

> [Averaging Quaternions](https://github.com/tolgabirdal/averaging_quaternions)?

Your reference is correct ，this code can be optimized for higher accuracy

**Comment by SoapJN** [comment]
> [Averaging Quaternions](https://github.com/tolgabirdal/averaging_quaternions)?

Your reference is correct ，This code can be optimized for higher accuracy

**Comment by savuor** [comment]
Yes, here's a mistake, a quaternion should be of unit norm, it's expected to be unit norm. This can be changed to eigenvalue-based method in the article you mentioned.
Or, for simplicity, just divide resulting quaternion by its norm. This is popular approach (see [this](https://www.cs.utah.edu/~ladislav/kavan08geometric/kavan08geometric.pdf) as an example extended to rotation+translation, paragraph _4.1 Coping with antipodality_) and it works accurate enough for some tasks if quaternions are close to each other.

**Comment by savuor** [comment]
@SoapJN why to close this? Implementing a method from the article or simply normalize quaternions would be a good thing. Something needs to be done before closing the issue

**Comment by SoapJN** [comment]
> @SoapJN why to close this? Implementing a method from the article or simply normalize quaternions would be a good thing. Something needs to be done before closing the issue

The points you made in your comments are very useful. This is my first time take part in issues on Github. What should I do next? Thinks for your advice.

**Comment by savuor** [comment]
@SoapJN creating an issue is useful thing by itself.
If you want, you can create a pull request that fixes the issue, I'll review it and tell what to fix if anything's wrong.
We have a guide on [How to contribute](https://github.com/opencv/opencv/wiki/How_to_contribute)

**Comment by SoapJN** [comment]
> @SoapJN creating an issue is a useful thing by itself.
> If you want, you can create a pull request that fixes the issue, I'll review it and tell what to fix if anything's wrong.
> We have a guide on [How to contribute](https://github.com/opencv/opencv/wiki/How_to_contribute)

Thinks for your help, I will finish it as soon as possible.
